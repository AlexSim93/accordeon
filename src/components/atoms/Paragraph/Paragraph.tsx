import styled from 'styled-components';
import { color, typography, space } from 'styled-system';

export const Paragraph = styled.p`
  ${space}
  ${typography}
  ${color}
`;

Paragraph.displayName = 'Paragraph';
