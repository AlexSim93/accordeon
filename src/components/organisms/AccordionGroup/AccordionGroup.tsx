import React, { useState } from 'react';
import { Box } from 'reflexbox';
import { space, SpaceProps } from 'styled-system';
import styled from 'styled-components';
import { Accordion } from '../../molecules/Accordion';

type Props = {
  initiallyExpanded?: number;
  content: any;
  DescriptionComponent: any;
};

type AccordionRenderProps = {
  id: string;
  title: string;
  description: string;
};

const AccordionContainer = styled.div<SpaceProps>`
  ${space}
`;

export const AccordionGroup: React.FC<Props> = ({
  initiallyExpanded,
  content,
  DescriptionComponent,
}) => {
  const [expandedIndex, changeExpandedIndex] = useState(initiallyExpanded);
  return (
    <Box as="section">
      {Array.isArray(content)
        && content.map(
          ({ id, title, description }: AccordionRenderProps, index: number) => (
            <AccordionContainer key={id} mb="m">
              <Accordion
                title={title}
                expanded={index === expandedIndex}
                onToggle={(state: boolean) => changeExpandedIndex(state ? -1 : index)}
              >
                <DescriptionComponent>{description}</DescriptionComponent>
              </Accordion>
            </AccordionContainer>
          ),
        )}
    </Box>
  );
};
