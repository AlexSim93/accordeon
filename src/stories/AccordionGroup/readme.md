# AccordionGroup

Используется для создания группы аккордеонов. Одновременно может быть открыт только один аккордеон. При открытии аккордеона, все остальные будут закрыты

## Свойства:

#### AccordionGroup

| Свойство             | Тип TS                                            | required | default   | Описание                                                              |
| -------------------- | ------------------------------------------------- | -------- | --------- | --------------------------------------------------------------------- |
| initiallyExpanded    | number                                            | ✗        | -1        | индекс аккордеона, который следует раскрыть изначально                |
| forceExpanded        | number                                            | ✗        | -         | Пропса которая позволяет раскрыть аккордеон                           |
| content              | {id: string; title: string; description: string;} | ✓        | -         | Данные которые необходимо поместить в Accordion                       |
| DescriptionComponent | React.FunctionComponent                           | ✓        | Paragraph | Компонент в который будет обернуто поле description из пропсы content |
| titleKey             | string                                            | ✓        | -         | ключ, по которому можно взять данные из content для заголовка         |
| descriptionKey       | string                                            | ✓        | -         | ключ, по которому можно взять данные из content для описания          |

## Использование:

- Группа аккордеонов { initiallyExpanded=0, content=[
  { id: "1", title: "accordion 1", description: "description 1" },
  { id: "2", title: "accordion 2", description: "description 2" },
  ], DescriptionComponent=Paragraph, titleKey="title"
  descriptionKey="description"}

```js
<AccordionGroup
  titleKey="title"
  descriptionKey="description"
  content={[
    { id: "1", title: "accordion 1", description: "description 1" },
    { id: "2", title: "accordion 2", description: "description 2" },
  ]}
  initiallyExpanded={1}
  DescriptionComponent={Paragraph}
/>
```
