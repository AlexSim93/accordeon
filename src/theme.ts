const breakpoints = ['768px', '1440px'];

export default {
  breakpoints,
  fontSizes: {
    xxs: '10px',
    xs: '12px',
    s: '14px',
    m: '16px',
    ml: '20px',
    l: '24px',
    xl: '32px',
    xxl: '40px',
    xxxl: '56px',
  },
  space: { // s 8, m 16, l 24
    xxxxs: '4px',
    xxxs: '8px',
    xxs: '12px',
    xs: '16px',
    s: '24px',
    sm: '32px',
    m: '40px',
    ml: '48px',
    l: '56px',
    xl: '64px',
    xxl: '72px',
    xxxl: '140px',
    xxxxl: '200px',
  },
  mediaQueries: {
    small: `@media screen and (min-width: ${breakpoints[0]})`,
    medium: `@media screen and (min-width: ${breakpoints[1]})`,
  },
  colors: {
    white: '#FFFFFF', // title text
    blackPearlLighten48: 'rgba(8, 13, 38, 0.48)', // border
    medianGrey: '#A7A7C8', // description text
    darkGreyLighten16: 'rgba(103, 107, 139, 0.16)', // accordion bg
    neutralDarkGrey: '#C4C4C4', // icon
    darkBlue: '#080D26', // page bg
    dodgerBlue: '#564FFF', // button bg
  },
  fonts: {
    Montserrat: '"Montserrat", sans-serif',
  },
  radii: {
    s: '4px',
  },
};
