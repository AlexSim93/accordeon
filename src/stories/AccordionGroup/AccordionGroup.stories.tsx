import React, { useState } from 'react';
import { storiesOf } from '@storybook/react';
import { Box, Flex } from 'reflexbox';
import { ThemeProvider } from 'styled-components';
import { Text } from '../../components/atoms/Text';
import theme from '../../theme';
// @ts-ignore
import notes from './readme.md';
import { Paragraph } from '../../components/atoms/Paragraph';
import { AccordionGroup } from '../../components/organisms/AccordionGroup';

storiesOf('Accordion Group', module).add(
  'Accordion Group',
  () => {
    const [openedIndex, changeState] = useState(0);
    return (
      <ThemeProvider theme={theme}>
        <Flex flexDirection="column">
          <Text color="black" fontWeight="600" fontSize="xl">
            Accordion Group
          </Text>
          <Box mb={24}>
            <Text color="black" fontSize="l">
              Accordion Group
            </Text>
            <AccordionGroup
              content={[
                { id: '1', title: 'accordion 1', description: 'description 1' },
                { id: '2', title: 'accordion 2', description: 'description 2' },
              ]}
              descriptionComponent={Paragraph}
              titleKey="title"
              descriptionKey="description"
            />
          </Box>
          <Box mb={24}>
            <Text color="black" fontSize="l">
              Accordion Group initially expanded second accordion
            </Text>
            <AccordionGroup
              titleKey="title"
              descriptionKey="description"
              content={[
                { id: '1', title: 'accordion 1', description: 'description 1' },
                { id: '2', title: 'accordion 2', description: 'description 2' },
              ]}
              initiallyExpanded={1}
              descriptionComponent={Paragraph}
            />
          </Box>
          <Box mb={24}>
            <Text color="black" fontSize="l">
              Accordion Group with external control
            </Text>
            <Box py={16}>
              <button type="button" onClick={() => changeState(0)}>
                Expand 1
              </button>
              <button type="button" onClick={() => changeState(1)}>
                Expand 2
              </button>
              <button type="button" onClick={() => changeState(-1)}>
                Close
              </button>
            </Box>
            <AccordionGroup
              titleKey="title"
              descriptionKey="description"
              content={[
                { id: '1', title: 'accordion 1', description: 'description 1' },
                { id: '2', title: 'accordion 2', description: 'description 2' },
              ]}
              forceExpanded={openedIndex}
              descriptionComponent={Paragraph}
            />
          </Box>
        </Flex>
      </ThemeProvider>
    );
  },
  { notes },
);
