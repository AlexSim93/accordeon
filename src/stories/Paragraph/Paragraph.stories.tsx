import React from 'react';
import { storiesOf } from '@storybook/react';
import { Box, Flex } from 'reflexbox';
import { ThemeProvider } from 'styled-components';
import { Paragraph } from '../../components/atoms/Paragraph';
import notes from './readme.md';
import { Text } from '../../components/atoms/Text';
import theme from '../../theme';

storiesOf('Paragraph', module).add(
  'Paragraph',
  () => (
    <ThemeProvider theme={theme}>
      <Flex flexDirection="column">
        <Text color="black" fontWeight="600" fontSize="xl">
          Paragraph
        </Text>
        <Box mb={24}>
          <Text color="black" fontSize="l">
            Default paragraph
          </Text>
          <Paragraph>
            Lorem ipsum dolor sit amet, consectetur adipisicing elit. Corporis
            consectetur mollitia dicta voluptas soluta! Accusantium, doloremque
            neque, deleniti temporibus eius corporis veritatis doloribus quam
            dolorum, impeditFaperiam? Corrupti, animi eos. Lorem ipsum dolor sit
            amet, consectetur adipisicing elit. Corporis consectetur mollitia
            dicta voluptas soluta! Accusantium, doloremque neque, deleniti
            temporibus eius corporis veritatis doloribus quam dolorum,
            impeditFaperiam? Corrupti, animi eos. Lorem ipsum dolor sit amet,
            consectetur adipisicing elit. Corporis consectetur mollitia dicta
            voluptas soluta! Accusantium, doloremque neque, deleniti temporibus
            eius corporis veritatis doloribus quam dolorum, impeditFaperiam?
            Corrupti, animi eos. Lorem ipsum dolor sit amet, consectetur
            adipisicing elit. Corporis consectetur mollitia dicta voluptas
            soluta! Accusantium, doloremque neque, deleniti temporibus eius
            corporis veritatis doloribus quam dolorum, impeditFaperiam?
            Corrupti, animi eos.
          </Paragraph>
        </Box>
        <Box mb={24}>
          <Text color="black" fontSize="l">
            Paragraph fontSize large, color blackPearlLighten48
          </Text>
          <Paragraph color="blackPearlLighten48" fontSize="l">
            Lorem ipsum dolor sit amet, consectetur adipisicing elit. Corporis
            consectetur mollitia dicta voluptas soluta! Accusantium, doloremque
            neque, deleniti temporibus eius corporis veritatis doloribus quam
            dolorum, impedit aperiam? Corrupti, animi eos.Lorem ipsum dolor sit
            amet, consectetur adipisicing elit. Corporis consectetur mollitia
            dicta voluptas soluta! Accusantium, doloremque neque, deleniti
            temporibus eius corporis veritatis doloribus quam dolorum,
            impeditFaperiam? Corrupti, animi eos.Lorem ipsum dolor sit amet,
            consectetur adipisicing elit. Corporis consectetur mollitia dicta
            voluptas soluta! Accusantium, doloremque neque, deleniti temporibus
            eius corporis veritatis doloribus quam dolorum, impeditFaperiam?
            Corrupti, animi eos.Lorem ipsum dolor sit amet, consectetur
            adipisicing elit. Corporis consectetur mollitia dicta voluptas
            soluta! Accusantium, doloremque neque, deleniti temporibus eius
            corporis veritatis doloribus quam dolorum, impeditFaperiam?
            Corrupti, animi eos.
          </Paragraph>
        </Box>
      </Flex>
    </ThemeProvider>
  ),
  { notes },
);
