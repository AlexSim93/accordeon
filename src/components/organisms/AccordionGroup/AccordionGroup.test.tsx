import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';
import 'jest-styled-components';
import { AccordionGroup } from './AccordionGroup';
import { Paragraph } from '../../atoms/Paragraph';
import theme from '../../../theme';
import { ThemeProvider } from 'styled-components';

describe('AccordionGroup', () => {
  it('AccordionGroup should render correctly by default', () => {
    const accordionGroup = renderer
      .create(
        <ThemeProvider theme={theme}>
          <AccordionGroup
            titleKey="title"
            descriptionKey="description"
            content={[
              { id: '1', title: 'accordion 1', description: 'description 1' },
              { id: '2', title: 'accordion 2', description: 'description 2' },
            ]}
          />
        </ThemeProvider>,
      )
      .toJSON();
    expect(accordionGroup).toMatchSnapshot();
  });

  it('AccordionGroup should render correctly when initiallyExpanded=1', () => {
    const accordionGroup = renderer
      .create(
        <ThemeProvider theme={theme}>
          <AccordionGroup
            titleKey="title"
            descriptionKey="description"
            content={[
              { id: '1', title: 'accordion 1', description: 'description 1' },
              { id: '2', title: 'accordion 2', description: 'description 2' },
            ]}
            initiallyExpanded={1}
            descriptionComponent={Paragraph}
          />
        </ThemeProvider>,
      )
      .toJSON();
    expect(accordionGroup).toMatchSnapshot();
  });

  it('AccordionGroup should render correctly when forceExpanded=1', () => {
    const accordionGroup = renderer
      .create(
        <ThemeProvider theme={theme}>
          <AccordionGroup
            titleKey="title"
            descriptionKey="description"
            content={[
              { id: '1', title: 'accordion 1', description: 'description 1' },
              { id: '2', title: 'accordion 2', description: 'description 2' },
            ]}
            forceExpanded={1}
            descriptionComponent={Paragraph}
          />
        </ThemeProvider>,
      )
      .toJSON();
    expect(accordionGroup).toMatchSnapshot();
  });

  it('Accordion should expand and collapse on click', () => {
    const accordionGroup = mount(
      <AccordionGroup
        titleKey="title"
        descriptionKey="description"
        content={[
          { id: '1', title: 'accordion 1', description: 'description 1' },
          { id: '2', title: 'accordion 2', description: 'description 2' },
        ]}
        descriptionComponent={Paragraph}
      />,
    );

    const accordionTitle = accordionGroup.find('TitleContainer').first();

    accordionTitle.simulate('click');

    expect(accordionGroup.find('DescriptionContainer')).toHaveLength(1);

    accordionTitle.simulate('click');

    expect(accordionGroup.find('DescriptionContainer')).toHaveLength(0);
  });
});
