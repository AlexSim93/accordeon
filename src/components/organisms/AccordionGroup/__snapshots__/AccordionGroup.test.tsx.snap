// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AccordionGroup AccordionGroup should render correctly by default 1`] = `
.c3 {
  font-size: s;
  font-family: Montserrat;
  color: white;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
}

.c1 {
  background-color: darkGreyLighten16;
  font-family: Montserrat;
  border-radius: s;
  font-style: normal;
}

.c4 {
  -webkit-transform: rotate( 360deg );
  -ms-transform: rotate( 360deg );
  transform: rotate( 360deg );
  -webkit-transition: -webkit-transform 0.2s;
  -webkit-transition: transform 0.2s;
  transition: transform 0.2s;
}

.c2 {
  padding: m;
  font-weight: 600;
}

.c2:hover {
  cursor: pointer;
}

.c0 {
  margin-bottom: m;
}

@media screen and (min-width:40em) {
  .c3 {
    font-size: m;
  }
}

@media screen and (min-width:52em) {
  .c3 {
    font-size: m;
  }
}

<section
  className="css-vurnku"
>
  <div
    className="c0"
  >
    <div
      className="c1 css-nwbphr"
    >
      <div
        className="c2 css-vpgf7p"
        onClick={[Function]}
      >
        <span
          className="c3"
          color="white"
          fontFamily="Montserrat"
          fontSize={
            Array [
              "s",
              "m",
              "m",
            ]
          }
        >
          accordion 1
        </span>
        <div
          className="c4 css-vurnku"
        />
      </div>
    </div>
  </div>
  <div
    className="c0"
  >
    <div
      className="c1 css-nwbphr"
    >
      <div
        className="c2 css-vpgf7p"
        onClick={[Function]}
      >
        <span
          className="c3"
          color="white"
          fontFamily="Montserrat"
          fontSize={
            Array [
              "s",
              "m",
              "m",
            ]
          }
        >
          accordion 2
        </span>
        <div
          className="c4 css-vurnku"
        />
      </div>
    </div>
  </div>
</section>
`;

exports[`AccordionGroup AccordionGroup should render correctly when forceExpanded=1 1`] = `
.c6 {
  height: 1px;
  background-color: blackPearlLighten48;
}

.c3 {
  font-size: s;
  font-family: Montserrat;
  color: white;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
}

.c1 {
  background-color: darkGreyLighten16;
  font-family: Montserrat;
  border-radius: s;
  font-style: normal;
}

.c4 {
  -webkit-transform: rotate( 360deg );
  -ms-transform: rotate( 360deg );
  transform: rotate( 360deg );
  -webkit-transition: -webkit-transform 0.2s;
  -webkit-transition: transform 0.2s;
  transition: transform 0.2s;
}

.c5 {
  -webkit-transform: rotate( 180deg );
  -ms-transform: rotate( 180deg );
  transform: rotate( 180deg );
  -webkit-transition: -webkit-transform 0.2s;
  -webkit-transition: transform 0.2s;
  transition: transform 0.2s;
}

.c2 {
  padding: m;
  font-weight: 600;
}

.c2:hover {
  cursor: pointer;
}

.c7 {
  padding-left: m;
  padding-right: m;
  padding-top: m;
  padding-bottom: m;
  color: medianGrey;
  font-weight: 500;
}

.c0 {
  margin-bottom: m;
}

.c8 {
  font-family: Montserrat;
  color: white;
}

@media screen and (min-width:40em) {
  .c3 {
    font-size: m;
  }
}

@media screen and (min-width:52em) {
  .c3 {
    font-size: m;
  }
}

@media screen and (min-width:40em) {
  .c7 {
    padding-bottom: l;
  }
}

@media screen and (min-width:52em) {
  .c7 {
    padding-bottom: l;
  }
}

<section
  className="css-vurnku"
>
  <div
    className="c0"
  >
    <div
      className="c1 css-nwbphr"
    >
      <div
        className="c2 css-vpgf7p"
        onClick={[Function]}
      >
        <span
          className="c3"
          color="white"
          fontFamily="Montserrat"
          fontSize={
            Array [
              "s",
              "m",
              "m",
            ]
          }
        >
          accordion 1
        </span>
        <div
          className="c4 css-vurnku"
        />
      </div>
    </div>
  </div>
  <div
    className="c0"
  >
    <div
      className="c1 css-nwbphr"
    >
      <div
        className="c2 css-vpgf7p"
        onClick={[Function]}
      >
        <span
          className="c3"
          color="white"
          fontFamily="Montserrat"
          fontSize={
            Array [
              "s",
              "m",
              "m",
            ]
          }
        >
          accordion 2
        </span>
        <div
          className="c5 css-vurnku"
        />
      </div>
      <div
        className="c6"
        height={1}
      />
      <div
        className="c7 css-th0e4m"
      >
        <p
          className="c8"
          color="white"
          fontFamily="Montserrat"
        >
          description 2
        </p>
      </div>
    </div>
  </div>
</section>
`;

exports[`AccordionGroup AccordionGroup should render correctly when initiallyExpanded=1 1`] = `
.c6 {
  height: 1px;
  background-color: blackPearlLighten48;
}

.c3 {
  font-size: s;
  font-family: Montserrat;
  color: white;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
}

.c1 {
  background-color: darkGreyLighten16;
  font-family: Montserrat;
  border-radius: s;
  font-style: normal;
}

.c4 {
  -webkit-transform: rotate( 360deg );
  -ms-transform: rotate( 360deg );
  transform: rotate( 360deg );
  -webkit-transition: -webkit-transform 0.2s;
  -webkit-transition: transform 0.2s;
  transition: transform 0.2s;
}

.c5 {
  -webkit-transform: rotate( 180deg );
  -ms-transform: rotate( 180deg );
  transform: rotate( 180deg );
  -webkit-transition: -webkit-transform 0.2s;
  -webkit-transition: transform 0.2s;
  transition: transform 0.2s;
}

.c2 {
  padding: m;
  font-weight: 600;
}

.c2:hover {
  cursor: pointer;
}

.c7 {
  padding-left: m;
  padding-right: m;
  padding-top: m;
  padding-bottom: m;
  color: medianGrey;
  font-weight: 500;
}

.c0 {
  margin-bottom: m;
}

.c8 {
  font-family: Montserrat;
  color: white;
}

@media screen and (min-width:40em) {
  .c3 {
    font-size: m;
  }
}

@media screen and (min-width:52em) {
  .c3 {
    font-size: m;
  }
}

@media screen and (min-width:40em) {
  .c7 {
    padding-bottom: l;
  }
}

@media screen and (min-width:52em) {
  .c7 {
    padding-bottom: l;
  }
}

<section
  className="css-vurnku"
>
  <div
    className="c0"
  >
    <div
      className="c1 css-nwbphr"
    >
      <div
        className="c2 css-vpgf7p"
        onClick={[Function]}
      >
        <span
          className="c3"
          color="white"
          fontFamily="Montserrat"
          fontSize={
            Array [
              "s",
              "m",
              "m",
            ]
          }
        >
          accordion 1
        </span>
        <div
          className="c4 css-vurnku"
        />
      </div>
    </div>
  </div>
  <div
    className="c0"
  >
    <div
      className="c1 css-nwbphr"
    >
      <div
        className="c2 css-vpgf7p"
        onClick={[Function]}
      >
        <span
          className="c3"
          color="white"
          fontFamily="Montserrat"
          fontSize={
            Array [
              "s",
              "m",
              "m",
            ]
          }
        >
          accordion 2
        </span>
        <div
          className="c5 css-vurnku"
        />
      </div>
      <div
        className="c6"
        height={1}
      />
      <div
        className="c7 css-th0e4m"
      >
        <p
          className="c8"
          color="white"
          fontFamily="Montserrat"
        >
          description 2
        </p>
      </div>
    </div>
  </div>
</section>
`;
